### PR TITLE
refactor(combineLatest): Improve types

### PR DESCRIPTION
**Pull request by benlesh**
- Reduces some type complexity in combineLatest
- Improves types around resultSelectors

### CODE REVIEW

File: spec-dtslint/observables/combineLatest-spec.ts [hunk 1]
```diff
@@ -37,31 +37,31 @@ it('should accept union types', () => {
 });
 
 it('should accept 1 param and a result selector', () => {
-  const o = combineLatest(a$, () => new A()); // $ExpectType Observable<A>
+  const o = combineLatest(a$, (a) => new A()); // $ExpectType Observable<A>
```
Comment (cartant) [inline spec-dtslint/observables/combineLatest-spec.ts:40]: nitpick: Specifying the parameter here does nothing, ATM, but we could improve the tests by using them:

```ts
const o = combineLatest(a$, (a) => [a] as const); // $ExpectType Observable<[A]>
```

File: src/internal/observable/combineLatest.ts [hunk 2]
```diff
@@ -11,322 +11,45 @@ import { popResultSelector, popScheduler } from '../util/args';
...
-  O1 extends ObservableInput<any>,
-  O2 extends ObservableInput<any>,
-  O3 extends ObservableInput<any>,
-  O4 extends ObservableInput<any>,
-  O5 extends ObservableInput<any>,
-  O6 extends ObservableInput<any>
->(
-  sources: [O1, O2, O3, O4, O5, O6],
+export function combineLatest<A extends readonly unknown[], R>(
+  sources: readonly [...ObservableInputTuple<A>],
```
Comment (cartant) [inline src/internal/observable/combineLatest.ts:17]: issue: `R` is not used in this sig. The return type should be `Observable<A>`, AFAICT.

nitpick: Why is the surrounding `[... ]` needed? Why not `sources: ObservableInputTuple<A>`? This occurs elsewhere, too.

File: src/internal/observable/combineLatest.ts [hunk 3]
```diff
@@ -11,322 +11,45 @@ import { popResultSelector, popScheduler } from '../util/args';
...
-export function combineLatest<R>(
-  ...observables: Array<ObservableInput<any> | ((...values: Array<any>) => R) | SchedulerLike>
+/** @deprecated Use the version that takes an array of Observables instead */
+export function combineLatest<A extends readonly unknown[], R>(...sources: [...ObservableInputTuple<A>, (...args: A) => R]): Observable<R>;
+/** @deprecated Use the version that takes an array of Observables instead */
+export function combineLatest<A extends readonly unknown[], R>(
+  ...sources: [...ObservableInputTuple<A>, (...args: A) => R, SchedulerLike]
 ): Observable<R>;
+/** @deprecated Use the version that takes an array of Observables instead */
+export function combineLatest<A extends readonly unknown[]>(...sources: [...ObservableInputTuple<A>, SchedulerLike]): Observable<A>;
```
Comment (cartant) [inline src/internal/observable/combineLatest.ts:44]: question: Why does this use an approach that differs from what's done with `merge`? Does this behave a little better? Are there tests to show this?

**Review by cartant — CHANGES_REQUESTED** [review]

File: src/internal/observable/combineLatest.ts (hunk 2 above)
Comment (cartant) [reply src/internal/observable/combineLatest.ts:17]: Having used this, I can see, now, that the `[... ]` business is needed.

Comment (kolodny) [inline src/internal/observable/combineLatest.ts:17]: OOO is `readonly` needed here? Doesn't make a difference IMO, purely curious

**Review by kolodny — APPROVED** [review]
I'm loving these PRs. I just feel for the Google folks who need to sync this in. Anyone with explicit generics is gonna have a bad time.

🔥🔥🔥🔥🔥🔥🔥🔥🔥 
image

**Comment by cartant** [comment]
Closing this 'cause the `combineLatest` types have been refactored in:

- https://github.com/ReactiveX/rxjs/pull/6092
- https://github.com/ReactiveX/rxjs/pull/6103

**Pull request closed without merging**